### ISSUE TITLE
Kubeconfig rotation does not work as expected

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener#3019]
**How to categorize this issue?**

/area usability
/kind bug
/priority critical

**What happened**:
After the rotation of the kubeconfig of a shoot the new kubeconfig is (sometimes?) not synced into the project namespace in the garden cluster, hence, the downloadable kubeconfig does not work and the end-user cannot interact with his cluster.

**What you expected to happen**:
The kubeconfig in the project namespace is up to date.

**How to reproduce it (as minimally and precisely as possible)**:
n/a yet, sorry

**Anything else we need to know?**:
I've seen occasions in which even the `kubecfg` secret in the shoot namespace in the seed was not correct, i.e., in those cases not even a second reconciliation helps, but only trying to rotate the credentials again.

**Environment**:

- Gardener version: v1.11.3

**Comment by vpnachev** [comment]
After some investigation I found several weird things
- the related secrets in the shoot namespace in the seed are re-generated
- the shootstate in the garden is never updated and keeps the old static-token
- sometimes the kubeconfig synced to the garden clusters is containing the deleted token, sometimes it is properly updated
- not every trigger of kubeconfig rotation reproduces the issue. 

Just wild guessing, it might be related to https://github.com/gardener/gardener/pull/2798 and to the cached clients, but I have no proofs about it. I will keep looking into it trying to understand more what exactly is happening.

**Comment by vpnachev** [comment]
I can confirm that the shoot state is never updated and this is always reproducible.
The reason seems to be the [reflect.DeepEqual](https://github.com/gardener/gardener/blob/2cfba7d68e6e8815f59b9620883afbe7f328c601/pkg/operation/botanist/secrets.go#L90) that does not detect differences correctly.

I have added some logs:
<details>


```diff
diff --git a/pkg/apis/core/v1alpha1/helper/shootstate_list.go b/pkg/apis/core/v1alpha1/helper/shootstate_list.go
index bc60271bf..dde5b29ec 100644
--- a/pkg/apis/core/v1alpha1/helper/shootstate_list.go
+++ b/pkg/apis/core/v1alpha1/helper/shootstate_list.go
@@ -85,6 +85,15 @@ func (g *GardenerResourceDataList) Get(name string) *gardencorev1alpha1.Gardener
 	return nil
 }
 
+// Keys returns the names of the items in the list
+func (g *GardenerResourceDataList) Keys() []string {
+	keys := []string{}
+	for _, resourceDataEntry := range *g {
+		keys = append(keys, resourceDataEntry.Name)
+	}
+	return keys
+}
+
 // Upsert inserts a new element or updates an existing one
 func (g *GardenerResourceDataList) Upsert(data *gardencorev1alpha1.GardenerResourceData) {
 	for i, obj := range *g {
diff --git a/pkg/operation/botanist/secrets.go b/pkg/operation/botanist/secrets.go
index 624fd2dfe..cab4fbe14 100644
--- a/pkg/operation/botanist/secrets.go
+++ b/pkg/operation/botanist/secrets.go
@@ -47,9 +47,12 @@ func (b *Botanist) GenerateAndSaveSecrets(ctx context.Context) error {
 	gardenerResourceDataList := gardencorev1alpha1helper.GardenerResourceDataList(b.ShootState.Spec.Gardener)
 
 	if val, ok := common.GetShootOperationAnnotation(b.Shoot.Info.Annotations); ok && val == common.ShootOperationRotateKubeconfigCredentials {
+		b.Logger.Info("[ROTATION] Will try to rotate secrets")
 		if err := b.rotateKubeconfigSecrets(ctx, &gardenerResourceDataList); err != nil {
+			b.Logger.Error("[ROTATION] Failed to rotate secrets")
 			return err
 		}
+		b.Logger.Info("[ROTATION] Secrets rotatet successfully")
 	}
 
 	if b.Shoot.Info.DeletionTimestamp == nil {
@@ -82,14 +85,18 @@ func (b *Botanist) GenerateAndSaveSecrets(ctx context.Context) error {
 	if shootWantsBasicAuth {
 		secretsManager = secretsManager.WithAPIServerBasicAuthConfig(basicAuthSecretAPIServer)
 	}
+	b.Logger.Infof("[ROTATION] ResourceList before generation: %v", secretsManager.GardenerResourceDataList.Keys())
 
 	if err := secretsManager.Generate(); err != nil {
 		return err
 	}
+	b.Logger.Infof("[ROTATION] ResourceList after generation: %v", secretsManager.GardenerResourceDataList.Keys())
 
 	if reflect.DeepEqual(secretsManager.GardenerResourceDataList, gardencorev1alpha1helper.GardenerResourceDataList(b.ShootState.Spec.Gardener)) {
+		b.Logger.Info("[ROTATION] Generated and existing state secrets are the same.")
 		return nil
 	}
+	b.Logger.Info("[ROTATION] Generated and existing state secrets differ.")
 	return b.SaveGardenerResourcesInShootState(ctx, secretsManager.GardenerResourceDataList)
 }
 
```

</details>


and find out that when there is a differences in the resources list, reflect does not detect it
```text
time="2020-10-27T21:30:01+02:00" level=info msg="[ROTATION] Will try to rotate secrets" operation=reconcile shoot=foo/bar
time="2020-10-27T21:30:01+02:00" level=info msg=Succeeded flow="Shoot cluster reconciliation" operation=reconcile shoot=foo/bar task="Generating etcd encryption configuration"
time="2020-10-27T21:30:02+02:00" level=info msg=Succeeded flow="Shoot cluster reconciliation" operation=reconcile shoot=foo/bar task="Deploying Kubernetes API server service in the Seed cluster"
time="2020-10-27T21:30:02+02:00" level=info msg="[ROTATION] Secrets rotatet successfully" operation=reconcile shoot=foo/bar
time="2020-10-27T21:30:02+02:00" level=info msg=Succeeded flow="Shoot cluster reconciliation" operation=reconcile shoot=foo/bar task="Waiting until Kubernetes API LoadBalancer in the Seed cluster has reported readiness"
time="2020-10-27T21:30:02+02:00" level=info msg="[ROTATION] ResourceList before generation: [ca ca-etcd ca-front-proxy ca-kubelet ca-metrics-server kube-apiserver kube-apiserver-kubelet kube-aggregator kube-controller-manager kube-controller-manager-server kube-scheduler kube-scheduler-server cluster-autoscaler gardener-resource-manager kube-proxy kube-state-metrics prometheus prometheus-kubelet gardener gardener-internal cloud-config-downloader monitoring-ingress-credentials monitoring-ingress-credentials-users ssh-keypair service-account-key etcd-server-cert etcd-client-tls metrics-server alertmanager-tls grafana-tls prometheus-tls dependency-watchdog-internal-probe dependency-watchdog-external-probe vpn-shoot vpn-seed vpn-seed-tlsauth vpa-tls-certs etcdEncryptionConfiguration]" operation=reconcile shoot=foo/bar
time="2020-10-27T21:30:02+02:00" level=info msg="[ROTATION] ResourceList after generation: [ca ca-etcd ca-front-proxy ca-kubelet ca-metrics-server kube-apiserver kube-apiserver-kubelet kube-aggregator kube-controller-manager kube-controller-manager-server kube-scheduler kube-scheduler-server cluster-autoscaler gardener-resource-manager kube-proxy kube-state-metrics prometheus prometheus-kubelet gardener gardener-internal cloud-config-downloader monitoring-ingress-credentials monitoring-ingress-credentials-users ssh-keypair service-account-key etcd-server-cert etcd-client-tls metrics-server alertmanager-tls grafana-tls prometheus-tls dependency-watchdog-internal-probe dependency-watchdog-external-probe vpn-shoot vpn-seed vpn-seed-tlsauth vpa-tls-certs etcdEncryptionConfiguration static-token]" operation=reconcile shoot=foo/bar
time="2020-10-27T21:30:02+02:00" level=info msg="[ROTATION] Generated and existing state secrets are the same." operation=reconcile shoot=foo/bar
```

For regular reconciliation there is no difference, which is expected
```
time="2020-10-27T21:40:47+02:00" level=info msg="[ROTATION] ResourceList before generation: [static-token ca ca-etcd ca-front-proxy ca-kubelet ca-metrics-server kube-apiserver kube-apiserver-kubelet kube-aggregator kube-controller-manager kube-controller-manager-server kube-scheduler kube-scheduler-server cluster-autoscaler gardener-resource-manager kube-proxy kube-state-metrics prometheus prometheus-kubelet gardener gardener-internal cloud-config-downloader monitoring-ingress-credentials monitoring-ingress-credentials-users ssh-keypair service-account-key etcd-server-cert etcd-client-tls metrics-server alertmanager-tls grafana-tls prometheus-tls dependency-watchdog-internal-probe dependency-watchdog-external-probe vpn-shoot vpn-seed vpn-seed-tlsauth vpa-tls-certs etcdEncryptionConfiguration]" operation=reconcile shoot=foo/bar
time="2020-10-27T21:40:47+02:00" level=info msg="[ROTATION] ResourceList after generation: [static-token ca ca-etcd ca-front-proxy ca-kubelet ca-metrics-server kube-apiserver kube-apiserver-kubelet kube-aggregator kube-controller-manager kube-controller-manager-server kube-schedulerkube-scheduler-server cluster-autoscaler gardener-resource-manager kube-proxy kube-state-metrics prometheus prometheus-kubelet gardener gardener-internal cloud-config-downloader monitoring-ingress-credentials monitoring-ingress-credentials-users ssh-keypair service-account-key etcd-server-cert etcd-client-tls metrics-server alertmanager-tls grafana-tls prometheus-tls dependency-watchdog-internal-probe dependency-watchdog-external-probe vpn-shoot vpn-seed vpn-seed-tlsauth vpa-tls-certs etcdEncryptionConfiguration]" operation=reconcile shoot=foo/bar
time="2020-10-27T21:40:47+02:00" level=info msg="[ROTATION] Generated and existing state secrets are the same." operation=reconcile shoot=foo/bar

```

**Comment by plkokanov** [comment]
Just a note: I've been trying to reproduce it with gardenlet from v1.12.0-dev-cf9a793c4cdce4a64cb329868bd0ebad7c31b8fd and haven't been able to for the past 40 minutes (I did around 10-12 rotations) 
How often does it happen with v1.11.3?

**Comment by vpnachev** [comment]
The code that syncs the credentials back to the Garden cluster is looking fine to me and I do not have any suspicious. 
I will keep on eye on 
> sometimes the kubeconfig synced to the garden clusters is containing the deleted token, sometimes it is properly updated

in the next weeks and if it is not reproduced, I will close the issue.

**Comment by ialidzhikov** [comment]
@vpnachev , do you have any observations on it, can it be closed now?
I tried to look into the test results for the integration test that you added. But after 20m of search for test results I found that the commit that adds the integration test was not cherry-picked into the release branch - only the fix commit. Ref https://github.com/gardener/gardener/commits/release-v1.12.

**Comment by vpnachev** [comment]
I faced the issue only once, but the shoot still was not reconciled by the fixed version. I would like observe it next week and if we do not see it, we can close this issue.

**Comment by vpnachev** [comment]
And, yes - the test was not cherry-picked as it is more like a feature. I was not sure whether we should do it or not.

**Comment by ialidzhikov** [comment]
Thank you for the update!

**Comment by ialidzhikov** [comment]
> I faced the issue only once, but the shoot still was not reconciled by the fixed version. I would like observe it next week and if we do not see it, we can close this issue.

Any new observations?

**Comment by vpnachev** [comment]
Nope. 
/close
fixed by #3069
